keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+items or selection via labels.
+
+Workloads make claims to request resources with claim parameters that use CRDs
+provided together with a DRA driver. As soon as such a custom CR is created, a
+CRD controller of the DRA driver must convert it into a in-tree claim parameter
+object that uses the "parameter types" defined by the semantic models.
+
+The scheduler receives both the node resource information and the claim
+parameters. It uses the in-tree claim parameters to identify and allocate
+suitable resources. It captures all information that the DRA driver kubelet
+plugin will need in the "allocation types" of the semantic models and stores
+that in the claim status. Then it schedules the pod.
+
+The DRA driver kubelet plugin receives that information when the pod is about to
+start. It checks that the hardware is still available and healthy, then
+configures it for the pod as specified in the allocation types.
+
+### Publishing node resources
+
+A semantic model defines an attributes type. This is a Go type that gets
+embedded inside the resource.k8s.io API, not a separate API kind. The kubelet
+queries the registered DRA driver kubelet plugins for resource information
+through a new gRPC call. A streaming gRPC API allows drivers to update the
+capacity information.
+
+For example, a simple counter could be described in a `counter` Go package as:
+```
+type Attributes struct {
+    // Some meta information just for logging and debugging. Not used for counting.
+	metav1.ObjectMeta
+
+	Count int64
+}
+```
+
+The kubelet then publishes this information in a new NodeResources object on
+behalf of the driver. This implies that kubelet, DRA driver, and apiserver must
+all support the same semantic models and the same versions of them. It might be
+possible to support version skew (= keeping kubelet at an older version than
+the control plane and the DRA drivers) in the future, but currently this is out
+of scope.
+
+### Using semantic parameters as claim parameters
+
+The semantic model also defines a type for claim parameters. For the example
+above, that could be:
+```
+type Parameters struct {
+	Required int64
+}
+```
+
+The `Parameters.Required` value gets compared against `Attributes.Count`.
+
+Suppose a vendor has a CRD which allows requesting a certain number of `frobnicators`:
+
+```
+kind: ClaimParameters
+apiVersion: dra.frobincatorInc.example.com/v1alpha1
+metadata:
+  name: my-claim-parameters
+  uid: abcd
+  resourceVersion: 123
+spec:
+  requiredFrobincators: 100
+```
+
+The vendor CRD controller needs to sync this CRs into one in-tree object:
+```
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+metadata:
+  name: frobnicator-claim-parameters
+  namespace: default
+generatedFrom:
+  kind: ClaimParameters
+  apiGroup: dra.frobincatorInc.example.com
+  name: my-claim-parameters
+  resourceVersion: 123
+  uid: abcd
+parameters:
+  - counter:
+      required: 100
```

Comment:
In this example it's assumed that the driver only manages "frobnicators".

A slightly better example and probably what the "counter" model should support is counting things by name:

```
parameters:
  - counter:
      name: frobincators
      required: 100
```

But I digress :stuck_out_tongue_winking_eye:  This is a discussion that we need to have over in https://github.com/kubernetes/enhancements/pull/4385.

Reply:
I can't keep the discussion straight across 3 KEPs - at this point the splitting that may once have had meaning is just confusing me.  I don't know what to do with #4385 right now

Reply:
discriminators are not universally used, but they do help with disambiguation

Reply:
Regarding 3 KEPs - agreed. I think we should just have two KEPs - one for main DRA, one for the general design of semantic models and the ones we want to deliver together. If we decide to only move *some* of the semantic models to beta and leave some in alpha, we would split those off at that time into their own KEP and featuregate (we don't have to move everything that was in alpha to beta at once, things can change between alpha and beta).

For discriminators...I sort of prefer not having them in this case...seems redundant. But I will leave it to the discretion of the api reviewer :)

Reply:
> I can't keep the discussion straight across 3 KEPs - at this point the splitting that may once have had meaning is just confusing me. I don't know what to do with https://github.com/kubernetes/enhancements/pull/4385 right now

How about this: we define the counter and selector model in this KEP and leave out mode switching? Otherwise this KEP is going to be blocked by all the discussion that I expect once we try to seriously define that more complex model.

A set of counters is simple enough that examples make sense just when looking at them.

> [union type needs a discriminator](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#automatic-resource-allocation-and-deallocation).

I don't understand how this link relates to union types. Other "one-of-many" APIs don't have any additional type information either, for example https://pkg.go.dev/k8s.io/api/core/v1#VolumeSource in https://pkg.go.dev/k8s.io/api/core/v1#Volume.

Lifting up the name might make sense.

Reply:
Maybe "mode switching" means something to me that is different than what you mean?

How can we think about moving forward with a model unless we don't have an idea of how we will accomodate MIG?